### PR TITLE
Waiting API

### PR DESCRIPTION
**Pull request by chenshay**

### CODE REVIEW

File: src/api/subscriptions.js [hunk 1]
```diff
@@ -134,6 +134,13 @@ export class Subscriptions {
    */
   setOnLinkComplete(callback) {}
 
+  /**
+   * TODO(chenshay): Figure out what type the promise returns.
+   * @param {Promise} accountPromise Publisher's promise to lookup account.
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: `!Promise`.

File: src/api/subscriptions.js [hunk 2]
```diff
@@ -134,6 +134,13 @@ export class Subscriptions {
    */
   setOnLinkComplete(callback) {}
 
+  /**
+   * TODO(chenshay): Figure out what type the promise returns.
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: If we return any result in the promise, I think it should simply be the passthrough value from the `accountPromise` for convenience. That way API clients won't need to overnest.

File: src/api/subscriptions.js [hunk 3]
```diff
@@ -208,6 +215,12 @@ export const SubscriptionFlows = {
 };
 
 
+/** @enum {string} */
+export const WaitingViews = {
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: What is this enum for?

File: src/runtime/waiting-api.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    /** @private @const {!ActivityIframeView} */
+    this.activityIframeView_ = new ActivityIframeView(
+        this.win_,
+        this.activityPorts_,
+        feUrl('/loginwaitingiframe'),
+        feArgs({
+          publicationId: deps.pageConfig().getPublicationId(),
+          productId: deps.pageConfig().getProductId(),
+          accountPromise: this.accountPromise_,
```
Comment (dvoytenko) [inline src/runtime/waiting-api.js]: We can't pass a promise here. Actually cannot pass any non-structured-clonable values.

File: src/runtime/waiting-api.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    this.openViewPromise_ = this.dialogManager_.openView(
+        this.activityIframeView_);
+
+    return this.accountPromise_.then(account => {
+      // Account was found.
+      this.dialogManager_.completeView(this.activityIframeView_);
+      return Promise.resolve(account);
+    }, () => {
+      this.dialogManager_.completeView(this.activityIframeView_);
+      return Promise.reject({
```
Comment (dvoytenko) [inline src/runtime/waiting-api.js]: Are we sure we want to automatically start this flow ourselves? There are a lot of configurations options there. I think it'd be best to allow a client code to call `deferredAccountCreation` if it wants to.

File: src/runtime/waiting-api.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    this.deps_.callbacks().triggerFlowStarted(
+        WaitingViews.LOGIN_WAITING_VIEW);
+
+    this.openViewPromise_ = this.dialogManager_.openView(
+        this.activityIframeView_);
+
+    return this.accountPromise_.then(account => {
+      // Account was found.
+      this.dialogManager_.completeView(this.activityIframeView_);
+      return Promise.resolve(account);
```
Comment (dvoytenko) [inline src/runtime/waiting-api.js]: `return account`?

File: src/api/subscriptions.js (hunk 2 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/api/subscriptions.js (hunk 1 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/api/subscriptions.js (hunk 3 above)
Comment (chenshay) [reply src/api/subscriptions.js]: no longer needed

File: src/runtime/waiting-api.js (hunk 4 above)
Comment (chenshay) [reply src/runtime/waiting-api.js]: done

File: src/runtime/waiting-api.js (hunk 6 above)
Comment (chenshay) [reply src/runtime/waiting-api.js]: done

File: src/runtime/waiting-api.js (hunk 5 above)
Comment (chenshay) [reply src/runtime/waiting-api.js]: done

File: src/api/subscriptions.js [hunk 7]
```diff
@@ -208,6 +214,10 @@ export const SubscriptionFlows = {
 };
 
 
+/** @type {string} */
+export const LOGIN_WAITING_VIEW = 'loginWaitingView';
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: I'd say we should either put it inside the "SubscriptionFlows" above or remove it completely. I don't think it's a useful analytical signal b/c publishers would always call this function themselves - i.e. they always aware when it's used.

File: src/runtime/runtime-test.js [hunk 8]
```diff
@@ -678,6 +677,15 @@ describes.realWin('Runtime', {}, env => {
       expect(stub).to.be.calledOnce.calledWithExactly(options, callback);
     });
 
+    it('should delegate "showWaitingIndicator"', () => {
+      configuredRuntimeMock.expects('showWaitingIndicator').once()
+          .returns(Promise.resolve());
```
Comment (dvoytenko) [inline src/runtime/runtime-test.js]: Like with the previous issue with return values: let's always test that the value returned here is the exact value returned by the configuredRuntime. An easy way is to return some random value here in the mock, e.g. `11` and then test that the result of `showWaitingIndicator.then(value => {...})` below is equal to the same value (`11`).

File: src/runtime/waiting-api.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    this.openViewPromise_ = this.dialogManager_.openView(
+        this.activityIframeView_);
+
+    return this.accountPromise_.then(account => {
+      // Account was found.
+      this.dialogManager_.completeView(this.activityIframeView_);
+      return account;
+    }, () => {
+      this.dialogManager_.completeView(this.activityIframeView_);
+      return Promise.reject('no account found');
```
Comment (dvoytenko) [inline src/runtime/waiting-api.js]: Let's rethrow the reason here - it'll be more obvious to the API user.

Comment (chenshay) [reply src/runtime/waiting-api.js]: done

File: src/runtime/runtime-test.js (hunk 8 above)
Comment (chenshay) [reply src/runtime/runtime-test.js]: done

File: src/api/subscriptions.js (hunk 7 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

**Pull request merged**